cpp/src/io/orc/writer_impl.cu

```diff
@@ -480,270 +520,276 @@ std::vector<Stream> writer::impl::gather_streams(orc_column_view *columns,
+  hostdevice_2dvector<gpu::StripeStream> *strm_desc)
 {
-  std::vector<StripeInformation> stripes(stripe_list.size());
-  size_t group        = 0;
-  size_t stripe_start = 0;
-  for (size_t s = 0; s < stripe_list.size(); s++) {
-    size_t stripe_group_end = group + stripe_list[s];
-
-    for (size_t i = 0; i < num_columns; i++) {
-      const auto *ck = &chunks[group * num_columns + i];
+  std::vector<StripeInformation> stripes(stripe_bounds.size());
+  for (auto const &stripe : stripe_bounds) {
+    for (size_t col_idx = 0; col_idx < enc_streams->size().first; col_idx++) {
+      const auto &strm = (*enc_streams)[col_idx][stripe.first];
 
       // Assign stream data of column data stream(s)
       for (int k = 0; k < gpu::CI_INDEX; k++) {
-        const auto stream_id = ck->strm_id[k];
+        const auto stream_id = strm.ids[k];
         if (stream_id != -1) {
-          auto *ss           = &strm_desc[s * num_data_streams + stream_id - num_index_streams];
+          auto *ss           = &(*strm_desc)[stripe.id][stream_id - num_index_streams];
           ss->stream_size    = 0;
-          ss->first_chunk_id = (group * num_columns + i);
-          ss->num_chunks     = (stripe_group_end - group);
-          ss->column_id      = i;
+          ss->first_chunk_id = stripe.first;
+          ss->num_chunks     = stripe.size;
+          ss->column_id      = col_idx;
           ss->stream_type    = k;
         }
       }
     }
 
-    group                   = stripe_group_end;
-    size_t stripe_end       = std::min(group * row_index_stride_, num_rows);
-    stripes[s].numberOfRows = stripe_end - stripe_start;
-    stripe_start            = stripe_end;
+    auto const stripe_group_end     = *stripe.cend();
+    auto const stripe_end           = std::min(stripe_group_end * row_index_stride_, num_rows);
+    stripes[stripe.id].numberOfRows = stripe_end - stripe.first * row_index_stride_;
   }
 
-  strm_desc.host_to_device(stream);
-  gpu::CompactOrcDataStreams(
-    strm_desc.device_ptr(), chunks.device_ptr(), strm_desc.size(), num_columns, stream);
-  strm_desc.device_to_host(stream);
-  chunks.device_to_host(stream, true);
+  strm_desc->host_to_device(stream);
+  gpu::CompactOrcDataStreams(*strm_desc, *enc_streams, stream);
+  strm_desc->device_to_host(stream);
+  enc_streams->device_to_host(stream, true);
 
   return stripes;
 }
 
 std::vector<std::vector<uint8_t>> writer::impl::gather_statistic_blobs(
-  orc_column_view const *columns,
-  size_t num_columns,
-  size_t num_rows,
-  size_t num_rowgroups,
-  std::vector<uint32_t> const &stripe_list,
-  std::vector<StripeInformation> const &stripes,
-  hostdevice_vector<gpu::EncChunk> &chunks)
+  host_span<orc_column_view const> columns, host_span<stripe_rowgroups const> stripe_bounds)
 {
-  size_t num_stat_blobs = (1 + stripe_list.size()) * num_columns;
-  size_t num_chunks     = chunks.size();
+  auto const num_rowgroups = *stripe_bounds.back().cend();
```

Comment:
`*(stripe_rowgroups_variable.cend())` evaluates to `first + size` which is the last row group in the stripe. Perhaps it can be made into a more readable and obvious function?

Reply:
yeah, something is off with this API, but I don't see how to improve it. The problem is that this is not quite the last row group, but last + 1.
I thought of making a class to hold the vector of `stripe_bound`s. It could have a size() member to get the total size of all stripes. Not sure if this is justified.

Reply:
Added a helper function to get the total stripes size. At least this use of cend() is now avoided.